# ctg-rnaseq

The script can not process samples that are run in different sequencing runs. These have top be processed separately.

## Requirements

1. Clone and build the Singularity container for this pipeline: (currently. https://github.com/cirrina/ctg-singularity/tree/main/rnaseq/v1.1).
  + Add the path to the .sif in the nextflow.config `container = ` parameter under process{}
 2. Make SURE that the `scriptsdir` and `singularity_container` hardcoded within 'rnaseq-primer' are valid and matches the correct version.
 2. Edit your samplesheet to fullill the requirements. See section `SampleSheet` below
 3. Edit the nextflow.config so that file paths are correct. Make sure that the STAR version installed in .sif matches the verrsion of STAR used to built references in nextflow.config.

## Important Folders & Variables that are hard-coded
+ rnaseq-primer
  + `scriptsdir` : Must match the script file name (version folder)
  + `project_root` : Location where project work folders are generated. This folder is intermittant and used only for analyses. Important files (delivery and ctg save) will be moved from this folder. This foldeer should be safe to delete upon a successful run. (Default location in wich delivery folder will be created default '/projects/fs1/shared/ctg-projects/rnaseq')
  + `delivery_root` : Location where delivery folder will be created '/projects/fs1/nas-sync/ctg-delivery/rnaseq'
  + `ctg_save_root` : Location where files to be archived by ctg will be saved  qc saved for ctg will be created
  + `projectdir` : generate project dir from input -i `project_id` parameter and `project_root`


## Run pipeline using Primer & Driver
.. 1-3 from above  
 4. Run the `naseq-primer`. This must be performed from *within the Illumina Sequencing Runfolder*. Note that the ctg-rnaseq scripts should not be added to $PATH but rather be executed from their location (thus allowing proper control of what verison that is executed). Shortly, the rnaseq-driver will perform the following:
  a. Generate a project work folder based on the `project_id` flag (-i) and the `project_root` parameter.
  b. Process the SampleSheet and generate two SampleSheets used for script (one for demux and one for data analysis).
    + illegal characters are removed.
    + Column settings are cross checked against database
    + Indexes are checked
    +
  c. Write run & project specific parameters: `nextflow.params.XXX`
 5. Optional: Modify parameters in the `nextflow.params.XXX` file. Individual settings can be  
 6. Run the `rnaseq-driver`. This must be performed from *within the Project directory generated by the primer*. The driver will be copied into the



```
sh /projects/fs1/shared/ctg-pipelines/ctg-rnaseq/v1.0/rnaseq-primer -i 2021_070 -s 2021_070_SampleSheet.csv
cd ..../ctg-projects/rnaseq/2021_070
cd

```

- This will use default values (e.g. /path/to/runfolder/CTG_SampleSheet.csv)

## Usage Seqonly-Driver
```
usage: prime-ctg-rnaseq [ -i projectid ] [ -s samplesheet ] [ -l force_lane ] [ -l force_replace_index ] [ -h help ]


Optional arguments:
project_id       -i : Set 'meta-id' for runfolder (e.g. 210330-seqonly). Default: Takes date of runfolder (before first _) and adds '-seqonly' as suffix
samplesheet   -s : Set samplesheet used for run (Default: CTG_SampleSheet.csv)
force_lane -b : String with bcl2fastq argument. e.g. '--minimum-trimmed-read-length 20 --mask-short-adapter-reads 20'
force_replace_index:        -r : Set if to resume nf-pipeline
help          -h : print help message
```

## Input

- Samplesheet (see `SampleSheet` section below)

### Pipeline steps:

* `Demultiplexing` (bcl2fastq): Converts raw basecalls to fastq, and demultiplex samples based on index (https://support.illumina.com/content/dam/illumina-support/documents/documentation/software_documentation/bcl2fastq/bcl2fastq2-v2-20-software-guide-15051736-03.pdf).
* `FastQC`: FastQC calculates quality metrics on raw sequencing reads (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). MultiQC summarizes FastQC reports into one document (https://multiqc.info/).
* `multiQC`: Compile fastQC and cellranger count metrics in multiqc report (https://multiqc.info/)
* `md5sum`: md5sum of all fastq files


### Output:
* ctg-PROJ_ID-output
    * `qc`: Quality control output.
        * fastqc output (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
        * multiqc output: Summarizing FastQC output and demultiplexing (https://multiqc.info/)
    * `fastq`: Contains raw fastq files from cellranger mkfastq.


### Samplesheet requirements:
Shpuld be in the foremat of Illumina IEM sample sheet.
The Bold/Italic field below must be correct!
Note that `Species` and `Pooled` are expected and must be added to a sheet generated by the IEM

[Header]
IEMFileVersion,5  
Investigator Name,X  
Experiment Name,X  
Date,YYYY-MM-DD  
Workflow,GenerateFASTQ  
Application,NovaSeq FASTQ Only  
Instrument Type,NovaSeq  
Assay,Nextera XT  
Index Adapters,"Nextera XT v2 Index Kit A"  
Chemistry,Amplicon  

[Reads]  
***26***  
***78***  

[Settings]  
Adapter,***CTGTCTCTTATACACATCT***  

[Data]  
Sample_ID,Sample_Name,Sample_Plate,Sample_Well,I7_Index_ID,index,I5_Index_ID,index2,Sample_Project,Description  
***S1***,***S1***,,,***N702***,***CGTACTAG***,,,2021_024,  
***S2***,***S2***,,,***N706***,***TAGGCATG***,,,2021_024,  

#### Optional Sample Sheet columns


### Container  
https://github.com/perllb/ctg-seqonly/tree/main/container  
