#!/usr/bin/env Rscript


## v1.0
## Rscript to check files generated by teh rnaseq pipeline
## Uses the ctg samplesheet as template
## Will check expected output files for all modules included in mextflow rnaseq pipeline
## Designed to work on
## 1. nf-output, i.e. temp folder
## 2. delivery and ctg output

## Output is a logfile with



# ========================================== ##
## LOAD PACKAGES                             ##
## ========================================== ##
library(optparse)
library(dplyr)


## ========================================== ##
## PARSE COMMAND-LINE PARAMETERS              ##
## ========================================== ##

option_list <- list(
  ## global params
  make_option(c("-i", "--project_id"     ), type="character" , default=NULL   , metavar="character"   , help="Project id. Required. NOTE: This will overwrite Sample_Project column in sample sheet"    ),
  make_option(c("-s", "--sample_sheet"  ), type="character" , default=NULL   , metavar="path"   , help="Sample Sheet, colData, where samples are rows and columns are sample annotations."    ),
  make_option(c("-f", "--fastq_dir"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),
  make_option(c("-s", "--star_dir"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),
  make_option(c("-a", "--fcounts_dir"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),
  make_option(c("-b", "--fastqscreen_dir"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),

  make_option(c("-e", "--fastqc_dir"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),
  make_option(c("-m", "--multiqc_dir"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),

  make_option(c("-d", "--delivery_folder"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),
  make_option(c("-c", "--ctg_folder"  ), type="character" , default=NULL   , metavar="path"   , help=""    ),

  ## helper cheklist files for checking input params and indexes

)

opt_parser <- OptionParser(option_list=option_list, add_help_option=FALSE)
opt        <- parse_args(opt_parser)

# ===================================================
#  File pre- and suffixes not supplied in sample sheet
# ===================================================
fastq.sufifx <- opt$fastq_suffix
bam.suffix <- opt$bam_suffix
index_columns <- c('Index_Plate_Well',	'I7_Index_ID',	'index',	'I5_Index_ID', 'index2')


# Samplesheet names - outputed from samplesheet rscript
samplesheet_demux="${projectdir}/SampleSheet-${projectid}-demux.csv"
samplesheet_ctg="${projectdir}/SampleSheet-${projectid}-ctg.csv"
samplesheet_original="${projectdir}/SampleSheet-${projectid}-original.csv"



# fastq files


# star


# star index


#
