#!/usr/bin/env Rscript


## v1.0
## Rscript to check files generated by teh rnaseq pipeline
## Uses the ctg samplesheet as template.
## Will check if expected output files from rnaseq pipeline are present
## Designed to work RECURSIVELYLY on a specified folder:
## 1. project work folder
## 2. delivery folder
## 3. ctg-qc archive folder.

## Output is a logfile with



# ========================================== ##
## LOAD PACKAGES                             ##
## ========================================== ##
library(optparse)
library(dplyr)


## ========================================== ##
## PARSE COMMAND-LINE PARAMETERS              ##
## ========================================== ##

option_list <- list(
  ## global params
  make_option(c("-i", "--project_id"     ), type="character" , default=NULL   , metavar="character"   , help="Project id. Required only for log purposes"    ),
  make_option(c("-s", "--sample_sheet"  ), type="character" , default=NULL   , metavar="path"   , help="Sample Sheet, colData, where samples are rows and columns are sample annotations."    ),
  make_option(c("-d", "--check_dir"     ), type="character" , default=NULL   , metavar="path"   , help="Project work dir"    ),
  make_option(c("-o", "--output"     ), type="character" , default='log.rscript.filecheck.csv'  , metavar="path"   , help='Name of output file'    )

)

opt_parser <- OptionParser(option_list=option_list, add_help_option=FALSE)
opt        <- parse_args(opt_parser)

cat("\n\n\n\n=============================================================================== \n   running file check  \n=============================================================================== \n\n")


## ---------------
## Debug stuff
## ---------------
debug.mode = FALSE
if(debug.mode){
  opt = list()
  opt$project_id = '2021_076'
  opt$sample_sheet= '~/ctg/ctg-projects/2021_076/samplesheets/SampleSheet-2021_076-ctg.csv'
  opt$check_dir= '~/ctg/ctg-projects/2021_076'
  opt$output <- 'log.rscript.filecheck.csv'
}



## ---------------
## Check input
## ---------------
logfile <- opt$output # logfile <- file.path(getwd(), 'log.rscript.filecheck.csv')
if(dirname(opt$output)==".") logfile <- file.path(getwd(), opt$output)



# ===================================================
#  File pre- and suffixes not supplied in sample sheet
# ===================================================
fastq.sufifx <- "_001.fastq.gz"
fastqc.sufifx <- "_001_fastqc.html"
bam.sufifx <- "_Aligned.sortedByCoord.out.bam"
star.complete.suffix <- "_Log.final.out"
bam.index.sufifx <- "_Aligned.sortedByCoord.out.bam.csi"
markdups.suffix <- "_bam.MarkDuplicates.metrics.txt"
fastqscren.suffix <- "_001_screen.txt"



# ===================================================
#  Parse sample sheet
# ===================================================
samplesheet <- read.delim(file=opt$sample_sheet, header = T, as.is = T, sep = ",")
my_samples <- samplesheet$Sample_ID
my_fastq <- samplesheet$fastq_1
if(any(samplesheet$Paired)) my_fastq <- c(my_fastq, samplesheet$fastq_2[samplesheet$Paired])
my_bam <- samplesheet$bam






# ===================================================
#  define checkfile function
# ===================================================
# will output a character vector wiht 3 slots.
 # 1. PASS or FAIL, if all files were found on at least 1 recursive dir from parent
 # 2. number of passed files. Indicate a partial FAIL
 # 3. what directory were files found. Will be NA if FAIL and 0 files found.

foo.checkfiles <- function(filenames, parent_dir){
  dirs <- list.dirs(path=parent_dir, full.names = T, recursive = T)
  temp_max <- c(0, dirs[1]) ## c(paste(0, length(filenames), sep = " of "), dirs[i])

  for(i in 1:length(dirs)){
    u <- sapply(file.path(dirs[i], filenames), file.exists)
    uu <- c(length(which(u)), length(which(!u)))

    if(all(u)){
      return(c('PASS',c(paste("ALL",uu[1],"expected files"),dirs[i])))
    }
    if(uu[1] > temp_max[1]) temp_max <- c(uu[1], dirs[i])  ## c(paste(uu[1], uu{2}, sep = " of "), dirs[i])
    next(i)
  }

  return(c('FAIL', paste(temp_max[1],"of",length(filenames),"expected file(s)"), temp_max[2]))
}


checklist_files = list()
checklist_files$project_id <- opt$project_id
checklist_files$input_samplesheet <- opt$sample_sheet


# ===================================================
#  Prepare output
# ===================================================

cat("## Rscript Logfile -- Check of output files.", file = logfile, sep = "\n", append = F)
cat("", file = logfile, sep = "\n", append = T)
cat("## Input params", file = logfile, sep = "\n", append = T)

cat(paste("Project", opt$project_id, sep = ","), file = logfile, sep = "\n", append = T)
cat(paste("Checked_parent_dir", opt$check_dir, sep = ","), file = logfile, sep = "\n", append = T)
cat(paste("SampleSheet", opt$sample_sheet, sep = ","), file = logfile, sep = "\n", append = T)

cat("## Rsctipt Chekfile results", file = logfile, sep = "\n", append = T)
cat("process,status,n_files,file_path", file = logfile, sep = "\n", append = T)


# ===================================================
#  step through all available processes & files
# ===================================================
# fastq files
my_files = my_fastq
x <- foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('fastq_files', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$fastq_files <- x[1]

# star_bam_files
my_files = my_bam
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('star_bam_files', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$star_bam_files <- x[1]

# star_completed
my_files = gsub(bam.sufifx, star.complete.suffix, my_bam)
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('star_completed', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$star_completed <- x[1]

# star_samtools_indexed
my_files = gsub(bam.sufifx, bam.index.sufifx, my_bam)
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('star_samtools_indexed', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$star_samtools_indexed <- x[1]

# picard_markdups
my_files = gsub(bam.sufifx, markdups.suffix, my_bam)
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('picard_markdups', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$picard_markdups <- x[1]

# fastqscreen
my_files = gsub(fastq.sufifx, fastqscren.suffix, my_fastq)
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('fastqscreen', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$fastqscreen <- x[1]

# fastqc (workdir)
# ================
my_files = gsub(fastq.sufifx, fastqc.sufifx, my_fastq)
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('fastqc_files', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$fastqc_files <- x[1]


# Other important files
# =====================

# multiqc
my_files <- paste0(opt$project_id, "_multiqc_report.html")
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('multiqc_files', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$multiqc_files <- x[1]

# logfiles


# samplesheets
my_files <- paste0("SampleSheet-",opt$project_id, "-demux.csv")
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('samplesheet_demux', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$samplesheet_demux <- x[1]

my_files <- paste0("SampleSheet-",opt$project_id, "-ctg.csv")
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('samplesheet_ctg', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$samplesheet_ctg <- x[1]

my_files <- paste0("SampleSheet-",opt$project_id, "-original.csv")
x <-foo.checkfiles(filenames = my_files, parent_dir= opt$check_dir)
cat(paste(c('samplesheet_original', x), collapse = ","), file = logfile, sep = "\n", append = T)
checklist_files$samplesheet_original <- x[1]

# Samplesheet names - outputed from samplesheet rscript
#samplesheet_demux="${projectdir}/SampleSheet-${projectid}-demux.csv"
#samplesheet_ctg="${projectdir}/SampleSheet-${projectid}-ctg.csv"
# samplesheet_original="${projectdir}/SampleSheet-${projectid}-original.csv"

cat("\n\n\n\n=============================================================================== \n   sample file check COMPLETE. Check log for appropriate actions  \n=============================================================================== \n\n")
print(checklist_files[1:2])
cat("\n=============================================================================== \n\n\n ")
print(checklist_files[unlist(checklist_files)=="PASS"])
cat("\n=============================================================================== \n\n\n ")
print(checklist_files[unlist(checklist_files)=="FAIL"])
cat("\n=============================================================================== \n\n\n ")
